README
$Id: README,v 1.26 2007/04/10 15:43:25 mpeppler Exp $
	
	   DBD::Sybase  --  a Sybase DBI driver for Perl 5.

   Copyright (c) 1996-2007  Michael Peppler

   You may distribute under the terms of either the GNU General Public
   License or the Artistic License, as specified in the Perl README file.


This Sybase DBI driver is built using the Sybase OpenClient Client
Library (aka CT-library). You will need to have that installed before
you can do anything with this package. You will also need Tim Bunce's
DBI module, latest release. DBI is available from CPAN, in

    "CPAN"/authors/id/TIMB/DBI-1.xx-tar.gz

The Sybase OpenClient libraries are of course available from Sybase.
See http://www.sybase.com or http://www.peppler.org for details.

An alternative is to use the FreeTDS reimplementation of Client Library.
With FreeTDS DBD::Sybase can be used to query a MS-SQL 7 or 2000
database server from a Unix/linux host.
See http://www.freetds.org for details.

If you use FreeTDS please read the README.freetds file.

DBD::Sybase is a reasonably complete implementation of the DBI spec, but 
there are still some features that are missing. Please see the DBD::Sybase
man page for details.

The full write-ups of the bugs that are listed in the CHANGES file as "fixed" 
can be seen at http://www.peppler.org/bugdb

Getting Help
============

If you are stuck please start by searching the Web, the appropriate
mailing lists and my bug database, as it is likely that someone else
will have experienced the same problem before you. Hint: Google is
your friend :-)

The sybperl-l mailing list is dedicated to Sybase <-> perl issues and 
covers both the sybperl modules (Sybase::CTlib, etc.) and DBD::Sybase.
It is archived and searchable at http://www.peppler.org/archive/sybperl-l/

If after seaching you still don't find any solution then post a question
(with as much detail as possible) to either dbi-users@perl.org,
sybperl-l@peppler.org or freetds@lists.ibiblio.org (only use the
latter if you use the FreeTDS libraries with DBD::Sybase).

For recent changes please see the CHANGES file.

Building:
---------

Make sure that the SYBASE environment variable points to the Sybase
installation that you want to use for this build. The easiest way
to do this is to source the SYBASE.sh/SYBASE.csh or SYBASE.bat file
in the root directory of the Sybase installation you want to use.

This version attempts to find out which libraries it needs by looking
at $SYBASE/lib ($SYBASE/$SYBASE_OCS/lib for ASE 12.x installations.) 
This behaviour works on Unix and VMS systems, but on Win32 you still
have to edit the CONFIG file the old way). 
It assumes that $SYBASE is set and points to 
the Sybase installation directory that you want to use to build 
DBD::Sybase, and will also use the value of $SYBASE_OCS if it is set.

Then run 
	perl Makefile.PL
You will be prompted to choose build options (chained mode for AutoCommit, 
threaded libraries when using a threaded perl) and for server/user/pwd
to use for the "make test" step.

If there were warnings about missing libraries, go to the manual 
configuration, below.

Run
	make
If the make failed, go to the manual configuration, below.
Run
	make test
If this succeeds you can install the package, via
	make install
(which you probably have to run as root.)

Manual Configuration:
---------------------

If the automatic configuration fails, then you will have to edit the
CONFIG file, and set the EXTRA_LIBS entry according to your OS and
Sybase release, and then run 
	perl Makefile.PL --file

The CONFIG file can be used to set certain system variables that
are needed for the build. 
	SYBASE is the root directory of your Sybase installation. DBD::Sybase
	will use $SYBASE/lib and $SYBASE/include during the build.

	EXTRA_LIBS lists any extra libraries that are required on your
	system. For example, Solaris 2.x needs -ltli. See your OS specific
	documentation supplement from Sybase to determine what is required.

	DBI_INCLUDE is the directory where DBI installed its include files.
	Makefile.PL will normally deduce this directory from perl's Config
	module, so you only need to set this if Makefile.PL fails.

	LINKTYPE. Uncomment and set to 'static' if you want to build
	DBD::Sybase statically (ie always included in a new perl binary).

Edit PWD, and set the user, password and server that you want to use for the 
'make test'.

Run perl Makefile.PL, make, make test.

If everything's fine, run "make install" to move the files to your 
installed perl library tree.

Automated build with no prompts:
--------------------------------

You can run Makefile.PL with command line arguments to accept defaults
and build DBD::Sybase in an automated manner (without prompts).

The syntax is:

perl Makefile.PL --accept_test_defaults --chained {Y, N} --threaded_libs {Y, N}



Known Problems:
===============

Solaris 2.x:
-----------
On Solaris 2.x make test will fail if LD_LIBRARY_PATH is set and has
/usr/lib or /lib before $SYBASE/lib. This is because both Solaris 2.x
and Sybase have a library called libintl.so, and if /usr/lib is placed 
before $SYBASE/lib in LD_LIBRARY_PATH the dynamic loader will search
the wrong library when loading DBD::Sybase. In general it is not
necessary to set LD_LIBRARY_PATH on Solaris, and it is only rarely
necessary to include /usr/lib or /lib in the LD_LIBRARY_PATH as those
directories will be searched by default.

Linux:
------
If the LANG or LC_ALL environment variable is set and points to an 
entry that does not exist in $SYBASE/locales/locales.dat (in the 
[linux] section) then you may get a core dump. This is an
OpenClient problem. 
See http://www.peppler.org/FAQ/linux.html#q1.14 for
additional details.

FreeTDS:
--------

See the README.freetds file for details.

Regression Tests:
-----------------

The regression tests cover the normal operations, but can't check
for all combination of data and access modes. You should always test
with your own scripts/data before moving a new release into production.

NOTE: t/fail.t will *fail* on test #8 if you are connecting to 
      an 11.0.3.3 server. This is a bug in the server, and not something
      that I can do anything about, unfortunately. It's a fairly
      obscure constraint violation condition - read the code if
      you are interested.

I have a simple bug tracking database at http://www.peppler.org/bugdb
You can use it to check for known problems, and to report new ones.

See perldoc DBD::Sybase for details about the package.

Comments, criticism, etc. welcome!

Michael
--
Michael Peppler
mpeppler@peppler.org
http://www.peppler.org/



CHANGES
$Id: CHANGES,v 1.88 2011/10/02 15:03:17 mpeppler Exp $

Release 1.14

    Fix bad size handling for unicode data. 
    Remove default charset setting to utf8 (this had been done in 1.11 as part of
    improved utf8 handling, but has negative side-effects. If unicode handling is needed
    then set "charset=utf8" as part of the connection string.
    Enforce the fact that utf8/unicode handling only works with OpenClient 15.x or later.

Release 1.13

    Fix for incorrect UTF8 handling when retrieving UNICODE data (Jean-Pierre Rupp).

Release 1.12

    Bug/Typo/Compatibility fixes with various versions of OpenClient.
    Experimental: Handle in/out parameters (Merijn Broeren)
    
Release 1.11

    Remove reliance on PERL_POLLUTE.
    Add better support for utf8 (Dave Rolsky)
    
Release 1.10

    Handle 15.x datatypes correctly.
    Add LONGMS date format symbol to handle microseconds for bigdatetime.
    Add support for CS_LONGCHAR_TYPE (Mark Aufflick)
    Document syb_isdead().
    Handle 64bit builds with FreeTDS (Ian Grant/Hans Kristian Rosbach)
    Add foreign_key_info & statistics_info (Jim Radford)
    Change behavior of large fixed precision numeric types (money, bigint)
    to be converted to a string internally and returned as such to the caller
    (behavior similar to numeric/decimal). This can be reverted to the old behavior 
    by defining SYB_NATIVE_NUM.

Release 1.09

    Behavior change: A new connection level attribute (syb_disconnect_in_child)
    has been added to allow automatical handling of InactiveDestroy across
    forks. By default in 1.09 a connection will NOT get closed if the process
    ID of the process that is exiting is not the same as the PID of the
    process that created the connection.
    Detect ASE error 5702 (The server is terminating this process) as a fatal
    error for the connection.
    
    Bug Fixes
    
    645 - Spurious COMMIT calls sent to the dataserver during the login/connect calls.
    628 - Increase size of Kerberos Principal string buffer
    627 - Spurious sigset_t declaration.

Release 1.08

	Detect missing libblk.a library, and disable the BLK api calls
	if necessary.
	Added code to force dlopen() to use RTLD_GLOBAL.
	Corrected ct_option() functionality detection.
	Fixed incorrect handling of bind_params() (Thanks to Tim Bunce).
	Added serverType DSN parameter.
	Added tds_keepalive DSN parameter.
	Fixed incorrect handling of multiple result sets with DBI
	1.53 and later.
	Re-wrote $dbh->ping() in C, it's now four times faster.
	Allow automated build without prompts.
	Improved nsql().
	Added corrected handling of DATE and TIME values (ASE 12.5.2 and later).
	Added handling of UNSIGNED INT and BIGINT (ASE 15 and later).
	Added PERL_NO_GET_CONTEXT #define.

        Bug Fixes
	
	624 - Empty strings incorrectly passed as NULL.
	616 - Spurious error message when the login request times out.
	614 - Documentation improvement for syb_xxx methods.
	610 - Segfault when using signals with the threaded libraries and
	      perl >= 5.8.


Release 1.07

	Changed the t/xblk.t test to lookup the charset used by the 
        server and specify this in the connect() string. This should
        avoid failures when the client and server uses charsets of
        different sizes (utf8 vs. iso_1, for example).
	Better error reporting when the connection data is incorrect 
	for the test scripts.
	Modified $dbh->ping() slightly.

        Bug Fixes

	604 - Add missing mode parameter to mkdir in t/xblob.t
	606 - Memory leak in the BLK API.

Release 1.06

	Fix off-by-one error for ISO date format.
	Clear error/warning when connecting to a Replication Server.
        Fix AutoCommit "off" behavior when CHAINED mode is turned off.
        Fix $dbh->begin_work() behavior.

        Note: This version fails 4 tests in t/xblk.t when building
        against the 15.0 Beta OCS libraries.
        
        Bug Fixes

	582 - ISO date formatting off by one for months.
	591 - NUM_OF_PARAMS isn't handled properly
        593 - Connection can become unusable due a bug in 
              get_server_version().
	597 - Prepared stored procs with placeholders return
	      corrupted recordset on second fetch.
        599 - The call to "prepare" also executes the statement.
        600 - $sth->finish sometimes fails to properly clean up the
              handle.

Release 1.05

        BEHAVIOR CHANGE - $dbh->{LongReadLen} must now be called
	before $dbh->prepare(). Previously you could call this after
	the $dbh->prepare() but before the $sth->execute().

        Install private statement handle methods for TEXT/IMAGE handling
        to avoid $h->func() calls, and update documentation.
        Implement experimental BLK API via prepare/execute loop.
	Change default "AutoCommit" off mode from explicit transactions
	to using the "chained" mode if it is available.
	Add $sth->syb_describe() call, taken from Sybase::CTlib's 
	ct_describe().
	Add ISO8601 date/time format for output.
	Fix $sth->finish() behavior when syb_flush_finish is turned on.
	Changed do { } while($sth->{syb_more_results}); idiom to use
	redo instead.
	Better/more consistent handling of multiple sth on a single dbh,
        and new test file.

	Bugs Fixed:

	580 - Binding binary/varbinary values to placeholders sometimes
	      fails.
	575 - Fails three tests under Tru-64.
	577 - perl Makefile.PL fails if umask is 0.
        578 - Better warning for calling $dbh->{LongReadLen} if $dbh is busy.
        572 - Minor documentation update for bind_param().

Release 1.04

	Bugs Fixed:

	566 - $sth->{NAME} fails right after prepare().

Release 1.03

	Added linking of threaded Sybase libs if perl is built with
	threading turned on.

Sybase.h
/* $Id: Sybase.h,v 1.19 2011/04/25 08:59:17 mpeppler Exp $

   Copyright (c) 1997 - 2011 Michael Peppler

   You may distribute under the terms of either the GNU General Public
   License or the Artistic License, as specified in the Perl README file.

*/



#define NEED_DBIXS_VERSION 93

#define PERL_NO_GET_CONTEXT

#include <DBIXS.h>		/* installed by the DBI module	*/
#include "dbivport.h"

#include <ctpublic.h>
#include <bkpublic.h>

/* These defines avoid name clashes for multiple statically linked DBD's    */
 
#define dbd_init        syb_init
#define dbd_db_login6       syb_db_login
#define dbd_db_do       syb_db_do
#define dbd_db_commit       syb_db_commit
#define dbd_db_rollback     syb_db_rollback
#define dbd_db_disconnect   syb_db_disconnect
#define dbd_discon_all   syb_discon_all
#define dbd_db_destroy      syb_db_destroy
#define dbd_db_STORE_attrib syb_db_STORE_attrib
#define dbd_db_FETCH_attrib syb_db_FETCH_attrib
#define dbd_st_prepare      syb_st_prepare
#define dbd_st_rows     syb_st_rows
#define dbd_st_execute      syb_st_execute
#define dbd_st_fetch        syb_st_fetch
#define dbd_st_finish       syb_st_finish
#define dbd_st_destroy      syb_st_destroy
#define dbd_st_blob_read    syb_st_blob_read
#define dbd_st_STORE_attrib syb_st_STORE_attrib
#define dbd_st_FETCH_attrib syb_st_FETCH_attrib
#define dbd_describe        syb_describe
#define dbd_bind_ph     syb_bind_ph



/* read in our implementation details */

#include "dbdimp.h"

#if defined(CS_CURRENT_VERSION)
#define CTLIB_VERSION	CS_CURRENT_VERSION
#else
#if defined(CS_VERSION_125)
#define CTLIB_VERSION   CS_VERSION_125
#else 
#if defined(CS_VERSION_120)
#define CTLIB_VERSION   CS_VERSION_120
#else 
#if defined(CS_VERSION_110)
#define CTLIB_VERSION   CS_VERSION_110
#else
#define CTLIB_VERSION	CS_VERSION_100
#endif
#endif
#endif
#endif

#if defined(CS_UNICHAR_TYPE)
#if defined (is_utf8_string)
#define DBD_CAN_HANDLE_UTF8
#endif
#endif

/*#define CTLIB_VERSION	CS_VERSION_100 */

#ifndef MAX
#define MAX(X,Y)	(((X) > (Y)) ? (X) : (Y))
#endif

#ifndef MIN
#define MIN(X,Y)	(((X) < (Y)) ? (X) : (Y))
#endif


#if !defined(Sybase_h)
#define Sybase_h 1

void     syb_init _((dbistate_t *dbistate));

int      syb_discon_all _((SV *drh, imp_drh_t *imp_drh));

int      syb_db_login _((SV *dbh, imp_dbh_t *imp_dbh, char *dbname, char *uid, char *pwd, SV *attribs));
int      syb_db_do _((SV *sv, char *statement));
int      syb_db_commit     _((SV *dbh, imp_dbh_t *imp_dbh));
int      syb_db_rollback   _((SV *dbh, imp_dbh_t *imp_dbh));
int      syb_db_disconnect _((SV *dbh, imp_dbh_t *imp_dbh));
void     syb_db_destroy    _((SV *dbh, imp_dbh_t *imp_dbh));
int      syb_db_STORE_attrib _((SV *dbh, imp_dbh_t *imp_dbh, SV *keysv, SV *valuesv));
SV      *syb_db_FETCH_attrib _((SV *dbh, imp_dbh_t *imp_dbh, SV *keysv));

int      syb_st_prepare _((SV *sth, imp_sth_t *imp_sth,
                char *statement, SV *attribs));
int      syb_st_rows    _((SV *sth, imp_sth_t *imp_sth));
int      syb_st_execute _((SV *sth, imp_sth_t *imp_sth));
AV      *syb_st_fetch   _((SV *sth, imp_sth_t *imp_sth));
int      syb_st_finish  _((SV *sth, imp_sth_t *imp_sth));
void     syb_st_destroy _((SV *sth, imp_sth_t *imp_sth));
int      syb_st_blob_read _((SV *sth, imp_sth_t *imp_sth,
                int field, long offset, long len, SV *destrv, long destoffset));
int      syb_ct_get_data _((SV *sth, imp_sth_t *imp_sth, 
			    int column, SV *bufrv, int buflen));
int      syb_ct_data_info _((SV *sth, imp_sth_t *imp_sth, int action, 
			     int column, SV *attr));
int      syb_ct_send_data _((SV *sth, imp_sth_t *imp_sth, char *buffer, 
			     int size));
int      syb_ct_prepare_send _((SV *sth, imp_sth_t *));
int      syb_ct_finish_send _((SV *sth, imp_sth_t *));
int      syb_st_STORE_attrib _((SV *sth, imp_sth_t *imp_sth, SV *keysv, SV *valuesv));
SV      *syb_st_FETCH_attrib _((SV *sth, imp_sth_t *imp_sth, SV *keysv));
 
int      syb_describe _((SV *sth, imp_sth_t *imp_sth));
int      syb_bind_ph  _((SV *sth, imp_sth_t *imp_sth,
                SV *param, SV *value, IV sql_type, SV *attribs,
				int is_inout, IV maxlen));


/* prototypes for module-specific functions */
int      syb_thread_enabled _((void));
int      syb_set_timeout _((int timeout));
int      syb_db_date_fmt _((SV *, imp_dbh_t *, char *));

SV *     syb_set_cslib_cb ( SV *cb);


#endif /* defined Sybase_h */

/* end of Sybase.h */


Makefile.PL
# $Id: Makefile.PL,v 1.47 2010/11/06 14:28:11 mpeppler Exp $

use ExtUtils::MakeMaker;

require 5.004;

use strict;

# If either of these aren't available on your system then you need to
# get them!
use DBI;
use DBI::DBD;
use Config;
use Getopt::Long;

use vars qw($SYBASE $inc_string $lib_string $LINKTYPE $written_pwd_file
  $newlibnames $libdir);

$LINKTYPE         = 'dynamic';
$written_pwd_file = 'PWD';
$libdir           = 'lib';

my $file;
my $chained;
my $threaded_libs;
my $accept_test_defaults;
GetOptions(
	'--file'                 => \$file,
	'--chained:s'            => \$chained,
	'--threaded_libs:s'      => \$threaded_libs,
	'--accept_test_defaults' => \$accept_test_defaults
);

select(STDOUT);
$| = 1;

configure();

configPwd();

my $lddlflags = $Config{lddlflags};
$lddlflags = "-L$SYBASE/$libdir $lddlflags" unless $^O eq 'VMS';
my $ldflags = $Config{ldflags};
$ldflags = "-L$SYBASE/$libdir $ldflags" unless $^O eq 'VMS';

WriteMakefile(
	'NAME'         => 'DBD::Sybase',
	LIBS           => [$lib_string],
	INC            => $inc_string,
	clean          => { FILES => "Sybase.xsi $written_pwd_file" },
	OBJECT         => '$(O_FILES)',
	'VERSION_FROM' => 'Sybase.pm',
	'LDDLFLAGS'    => $lddlflags,

	#	      'LDFLAGS'      => $ldflags,
	LINKTYPE => $LINKTYPE,
	(
		$^O eq 'VMS'
		? ( MAN3PODS => { 'Sybase.pm' => 'blib/man3/DBD_Sybase.3' } )
		: ( MAN3PODS => { 'Sybase.pm' => 'blib/man3/DBD::Sybase.3' } )
	),
	(
		$] >= 5.005
		? (
			ABSTRACT => 'DBI driver for Sybase datasources',
			AUTHOR   => 'Michael Peppler (mpeppler@peppler.org)'
		  )
		: ()
	),
	(
		     $] >= 5.005
		  && $^O eq 'MSWin32'
		  && $Config{archname} =~ /-object\b/i ? ( CAPI => 'TRUE' ) : ()
	),
	'dist' => { 'TARFLAGS' => 'cvf', 'COMPRESS' => 'gzip' },
	( $^O eq 'VMS' ? ( PM => 'Sybase.pm' ) : () )
);

sub MY::postamble {
	return dbd_postamble();
}

sub configure {
	my %attr;
	my ( $key, $val );

	my $sybase_dir = $ENV{SYBASE};

	if ( !$sybase_dir ) {

		# PR 517 - getpwnam() isn't portable.
		eval q{
	    $sybase_dir = (getpwnam('sybase'))[7];
	};
	}

	open( IN, "CONFIG" ) || die "Can't open CONFIG: $!";
	while (<IN>) {
		chomp;
		next if /^\s*\#/;
		next if /^\s*$/;

		( $key, $val ) = split( /\s*=\s*/, $_ );
		$key =~ s/\s//g;
		$val =~ s/\s*$//;

		$attr{$key} = $val;
	}

	if ( -d $sybase_dir ) {
		$SYBASE = $sybase_dir;
	}
	else {
		if ( $attr{SYBASE} && -d $attr{SYBASE} ) {
			$SYBASE = $attr{SYBASE};
		}
	}

	if ( !$SYBASE || $SYBASE =~ /^\s*$/ ) {
		die
"Please set SYBASE in CONFIG, or set the \$SYBASE environment variable";
	}

	$SYBASE = VMS::Filespec::unixify($SYBASE) if $^O eq 'VMS';
	if ( $^O eq 'darwin' ) {
		my $osxdir = 'Applications/Sybase/System';
		$SYBASE = $osxdir if -d $SYBASE && -d $osxdir;
	}

	# System 12.0 has a different directory structure...
	if ( defined( $ENV{SYBASE_OCS} ) ) {
		$SYBASE .= "/$ENV{SYBASE_OCS}";
	}

	my $libfound = 0;
	my $libsub   = '';
	foreach $libdir ( 'lib', 'lib64' ) {
		if ( -d "$SYBASE/$libdir" ) {
			if ( checkLib($SYBASE) ) {
				$libfound = 1;
				$libsub   = $libdir;
			}
		}
	}

	die "Can't find any Sybase libraries in $SYBASE/lib or $SYBASE/lib64"
	  unless $libfound;

	my $inc_found = 0;
	if ( -d "$SYBASE/include" && -f "$SYBASE/include/cspublic.h" ) {
		++$inc_found;
		$inc_string = "-I$SYBASE/include";
	}

	# In some freetds installations the include files have been moved
	# into /usr/include/freetds.
	if ( -d "$SYBASE/include/freetds" && "$SYBASE/include/freetds/cspublic.h" )
	{
		++$inc_found;
		$inc_string = "-I$SYBASE/include/freetds";
	}
	die "Can't find the Client Library include files under $SYBASE"
	  unless ($inc_found);

	my $version = getLibVersion("$SYBASE/$libsub");

	#    if(!$version || $version lt '12') {
	#print "FreeTDS or older Client Library. Enabling CS-Lib Callbacks\n";
	#$inc_string .= " -DUSE_CSLIB_CB=1";
	#    }

	checkChainedAutoCommit();

	#    print "OS = $^O\n";

	if ( $^O eq 'MSWin32' ) {
		$lib_string =
"-L$SYBASE/lib -llibct.lib -llibcs.lib -llibtcl.lib -llibcomn.lib -llibintl.lib -llibblk.lib $attr{EXTRA_LIBS} -lm";
	}
	elsif ( $^O eq 'VMS' ) {
		$lib_string =
"-L$SYBASE/lib -llibct.olb -llibcs.olb -llibtcl.olb -llibcomn.olb -llibintl.olb -llibblk.olb $attr{EXTRA_LIBS}";
	}
	elsif ( $^O =~ /cygwin/ ) {
		$lib_string = "-L$SYBASE/lib -lct -lcs -lblk";
		$inc_string .= " -D_MSC_VER=800";
	}
	else {

		# Supplied by Erick Calder. I'm not sure why libsybsrv is needed...
		$attr{EXTRA_LIBS} .= " -lsybsrv" if $^O eq 'darwin';

		my $extra = getExtraLibs( $SYBASE, $attr{EXTRA_LIBS}, $version );
		if ($file) {
			$lib_string =
"-L$SYBASE/$libdir -lct -lcs -ltcl -lcomn -lintl -lblk $attr{EXTRA_LIBS} -ldl -lm";
		}
		else {
			$lib_string =
"-L$SYBASE/$libdir -lct -lcs -ltcl -lcomn -lintl -lblk $extra -ldl -lm";
		}
		if ($newlibnames) {
			foreach (qw(ct cs tcl comn intl blk)) {
				$lib_string =~ s/-l$_/-lsyb$_/;
			}
		}
		elsif ( $^O =~ /linux|freebsd/i ) {
			$lib_string =~ s/-ltcl/-lsybtcl/;
		}

		my %libname;

		%libname = loadSybLibs("$SYBASE/$libdir");

		my $libtype = '';

		# Logic to replace normal libs with _r (re-entrant) libs if
		# usethreads is defined provided by W. Phillip Moore (wpm@ms.com)
		# I have no idea if this works on Win32 systems (probably not!)
		if ( $Config{usethreads} ) {
			print "Running in threaded mode - looking for _r libraries...\n";

			if ( checkForThreadedLibs() ) {
				my $found = 0;
				foreach ( split( /\s+/, $lib_string ) ) {
					next unless /^-l(\S+)/;
					my $oldname = $1;
					my $newname = $1 . "_r";
					next unless exists $libname{$newname};
					print "Found -l$newname for -l$oldname\n";
					++$found;
					$lib_string =~ s/-l$oldname\b/-l$newname/;
				}
				if ( !$found ) {
					print "No thread-safe Sybase libraries found\n";
					$inc_string .= ' -DNO_THREADS ';
				}
				else {
					$libtype .= '_r';
				}
			}
			else {
				print "OK - I'll use the normal libs\n\n";
			}
		}

		# If we are building for a 64 bit platform that also supports 32 bit
		# (i.e. Solaris 8, HP-UX11, etc) then we need to make some adjustments
		if ( $Config{use64bitall} ) {

			# Tru64/DEC OSF does NOT use the SYB_LP64 define
			# as it is ALWAYS in 64 bit mode.
			$inc_string .= ' -DSYB_LP64' unless $^O eq 'dec_osf';

			print "Running in 64bit mode - looking for '64' libraries...\n";

			my $found = 0;
			foreach ( split( /\s+/, $lib_string ) ) {
				next unless /^-l(\S+)/;
				my $oldname = $1;
				my $newname = $1 . '64';
				next unless exists $libname{$newname};
				print "Found -l$newname for -l$oldname\n";
				$lib_string =~ s/-l$oldname\b/-l$newname/;
				++$found;
			}
			if ($found) {
				$libtype .= '64';
			}
		}

		# Is the blk library available?
		#my @k = keys(%libname);
		#print "@k\n";
		#print "libtype = $libtype\n";
		if ( my @l = grep( /(syb)?blk$libtype/, keys(%libname) ) ) {
			print "BLK api available - found: @l\n";
		}
		else {
			print "BLK api NOT available.\n";
			$inc_string .= ' -DNO_BLK=1';
		}
	}

	my $config_sitearch  = $Config{sitearchexp};
	my $attr_dbi_include = $attr{DBI_INCLUDE};
	if ( $^O eq 'VMS' ) {
		$config_sitearch  = VMS::Filespec::unixify($config_sitearch);
		$attr_dbi_include = VMS::Filespec::unixify($attr_dbi_include);
	}
	my @try = ( @INC, $Config{sitearchexp} );
	unshift @try, $attr{DBI_INCLUDE} if $attr{DBI_INCLUDE};
	my $dbidir;
	for my $trydir (@try) {
		if ( -e "$trydir/auto/DBI/DBIXS.h" ) {
			$dbidir = "$trydir/auto/DBI";
			last;
		}
	}
	die "Can't find the DBI include files. Please set DBI_INCLUDE in CONFIG"
	  if !$dbidir;
	$inc_string .= " -I$dbidir";
	if ( $attr{LINKTYPE} ) {
		$LINKTYPE = $attr{LINKTYPE};
	}
}

sub loadSybLibs {
	my $dir = shift;

	my %libname = ();

	opendir( SYBLIB, $dir )
	  or die "Unable to opendir $dir: $!\n";
	foreach ( readdir(SYBLIB) ) {
		next unless -f "$dir/$_";
		next unless /^lib(\S+)\.(so|a|sl)/;
		$libname{$1} = 1;
	}

	closedir(SYBLIB);

	return %libname;
}

sub getLibVersion {
	my $lib = shift;

	opendir( DIR, $lib );

	# reverse to pick up libsybct before libct...
	my @files = reverse( grep( /lib(syb)?ct(64)?\./, readdir(DIR) ) );
	closedir(DIR);
	my $file;
	foreach (@files) {
		$file = "$lib/$_";
		last if -e $file;
	}

	open( IN, $file ) || die "Can't open $file: $!";
	binmode(IN);
	my $version;
	while (<IN>) {
		if (/Sybase Client-Library\/([^\/]+)\//) {
			$version = $1;
			last;
		}
	}
	close(IN);
	if ( !$version ) {
		print "Unknown Client Library version - assuming FreeTDS.\n";
	}
	else {
		print "Sybase OpenClient $version found.\n";
	}

	return $version;
}

sub getExtraLibs {
	my $dir     = shift;
	my $cfg     = shift;
	my $version = shift;

	opendir( DIR, "$dir/$libdir" ) || die "Can't access $dir/$libdir: $!";
	my %files =
	  map { $_ =~ s/lib([^\.]+)\..*/$1/; $_ => 1 }
	  grep( /lib/ && -f "$dir/$libdir/$_", readdir(DIR) );
	closedir(DIR);

	my %x = map { $_ => 1 } split( ' ', $cfg );
	my $dlext = $Config{dlext} || 'so';
	foreach my $f ( keys(%x) ) {
		my $file = $f;
		$file =~ s/-l//;
		next if ( $file =~ /^-/ );
		delete( $x{$f} ) unless exists( $files{$file} );
	}

	foreach my $f (qw(insck tli sdna dnet_stub tds skrb gss)) {
		$x{"-l$f"} = 1 if exists $files{$f} && -f "$dir/$libdir/lib$f.$dlext";
	}
	if ( $version gt '11' ) {
		delete( $x{-linsck} );
		delete( $x{-ltli} );
	}

	#    if($version ge '12.5.1') {
	#	delete($x{-lskrb});
	#    }

	join( ' ', keys(%x) );
}

sub checkLib {
	my $dir = shift;

	opendir( DIR, "$dir/$libdir" ) || die "Can't access $dir/$libdir: $!";
	my @files = grep( /libct|libsybct/i, readdir(DIR) );
	closedir(DIR);
	if ( grep( /libsybct/, @files ) ) {
		$newlibnames = 1;
	}
	else {
		$newlibnames = 0;
	}

	scalar(@files);
}

sub configPwd {
	open( IN, "PWD.factory" ) || die "Can't open PWD.factory: $!";
	my %pwd;
	while (<IN>) {
		chomp;
		next if (/^\s*\#/);
		next if (/^\s*$/);
		my ( $key, $val ) = split( /=/, $_ );
		$pwd{$key} = $val || "undef";
	}
	close(IN);

	if ($accept_test_defaults) {
		$pwd{SRV} = $pwd{SRV};
		$pwd{UID} = $pwd{UID};
		$pwd{PWD} = $pwd{PWD};
		$pwd{DB}  = $pwd{DB};
	}
	else {
		print
"The DBD::Sybase module need access to a Sybase server to run the tests.\n";
		print "To clear an entry please enter 'undef'\n";
		print "Sybase server to use (default: $pwd{SRV}): ";
		$pwd{SRV} = getAns(0) || $pwd{SRV};
		print "User ID to log in to Sybase (default: $pwd{UID}): ";
		$pwd{UID} = getAns(0) || $pwd{UID};
		print "Password (default: $pwd{PWD}): ";
		$pwd{PWD} = getAns(1) || $pwd{PWD};
		print "Sybase database to use on $pwd{SRV} (default: $pwd{DB}): ";
		$pwd{DB} = getAns(0) || $pwd{DB};
	}

	warn
"\n* Writing login information, including password, to file $written_pwd_file.\n\n";

	# Create the file non-readable by anyone else.
	my $old_umask;
	unless ( $^O =~ /MSWin32/ ) {
		$old_umask = umask(077);
		warn "cannot umask(): $!" unless defined($old_umask);
	}
	open( OUT, ">$written_pwd_file" ) || die "Can't open $written_pwd_file: $!";
	unless ( $^O =~ /MSWin32/ ) {
		umask($old_umask) != 077 && warn "strange return from umask()";
	}
	print OUT <<EOF;
# This file contains optional login id, passwd and server info for the test
# programs:
# You probably don't want to have it lying around after you've made
# sure that everything works OK.

EOF
	foreach ( keys %pwd ) {
		$pwd{$_} = '' if $pwd{$_} eq 'undef';
		print OUT "$_=$pwd{$_}\n";
	}
	close(OUT);
}

sub checkChainedAutoCommit {
	print <<EOF;

By default DBD::Sybase 1.05 and later use the 'CHAINED' mode (where available)
when 'AutoCommit' is turned off. Versions 1.04 and older instead managed
the transactions explicitly with a 'BEGIN TRAN' before the first DML
statement. Using the 'CHAINED' mode is preferable as it is the way that
Sybase implements AutoCommit handling for both its ODBC and JDBC drivers.

EOF
	print "Use 'CHAINED' mode by default (Y/N) [Y]: ";
	my $ans;
	if ( defined($chained) ) {
		$ans = $chained;
	}
	else {
		$ans = getAns(0);
	}
	if ( $ans and $ans !~ /^y/i ) {
		$inc_string .= " -DNO_CHAINED_TRAN=1";
	}
	print "\n";
}

sub checkForThreadedLibs {
	my $ret = 1;
	if ( $] >= 5.008 ) {
		$ret = 0;

		print <<EOF;

***NOTE***
There is an incompatibility between perl (5.8.x or later) built in threaded 
mode and Sybase's threaded libraries, which means that signals delivered to 
the perl process result in a segment violation.

I suggest building DBD::Sybase with the normal libraries in this case to get
reasonable behavior for signal handling.

EOF
		print "Use the threaded (lib..._r) libraries [N]: ";
		my $ans;

		if ( defined($threaded_libs) ) {
			$ans = $threaded_libs;
		}
		else {
			$ans = getAns(0);
		}
		if ( $ans and $ans =~ /^y/i ) {
			$ret = 1;
		}
		print "\n";
	}

	return $ret;
}

sub getAns {
	my $flag = shift;

	if ( $flag && -t ) {
		eval {
			require Term::ReadKey;
			Term::ReadKey::ReadMode('noecho');
		};
	}
	my $ans = <STDIN>;
	if ( $flag && -t ) {
		eval { Term::ReadKey::ReadMode('normal'); };
		print "\n";    # because newline from user wasn't echo'd
	}
	$ans =~ s/^\s+//;
	$ans =~ s/\s+$//;

	return $ans;
}
